draft-mouris-cfrg-mastic.md

```diff
@@ -205,11 +205,42 @@ into Mastic in {{vidpf}}. The appendix includes supplementary material:
+
+* `bool(val: Any) -> bool` converts a value `val` to a Boolean.
+* `byte(number: int) -> bytes` returns the encoding of the input `int` as a byte.
+* `bytearray(source: Union[str, bytes, bytearray, Iterable[int]]) -> bytearray` returns a mutable sequence of bytes.
+* `bytes(source: Union[str, bytes, bytearray, Iterable[int]]) -> bytes` returns a immutable sequence of bytes.
+* `cast(typ: Type[T], val: Any) -> T` casts a value `val` to a specific type `typ`
+* `concat(parts: list[bytes]) -> bytes` returns a concatenated byte string from a list of bytes.
+* `dst(usage: int) -> bytes` returns a string of bytes encoding the name of the VDAF (i.e., "mastic"), the version number, and the usage type.
+* `encode_public_share(public_share: list[CorrectionWord]) -> bytes` encodes a list of `CorrectionWord` items into bytes.
+* `Field(integer: int) -> Field` returns an integer represented as a field element. The value of integer MUST be non-negative and less than Field.MODULUS.
+* `Field.as_unsigned() -> int` returns the integer representation of a field element.
+* `Field.encode_vec(vec: list[Field]) -> bytes` encodes a vector of field elements `vec` as a byte string.
+* `Field.zeros(length: int) -> list[Self]` returns a vector of zeros of size `length`.
+* `flp.decide(verifier: list[F]) -> bool` is a deterministic decision algorithm of a verifier that takes as input a verifier message and outputs a Boolean indicating if the measurement from which it was generated is valid.
+* `flp.decode(output: list[F], num_measurements: int) -> AggResult` maps a sum of aggregate shares to an aggregate result.
+* `flp.encode(measurement: Measurement) -> list[F]` encodes a raw measurement as a vector of field elements. The return value MUST be of length `MEAS_LEN`.
+* `flp.prove(meas: list[F], prove_rand: list[F], joint_rand: list[F]) -> list[F]` is a prover's deterministic proof-generation algorithm. Its inputs are the encoded measurement, the "prover randomness" `prove_rand`, and the "joint randomness" `joint_rand`. The prover randomness is used only by the prover, but the joint randomness is shared by both the prover and verifier.
+* `flp.query(meas: list[F], proof: list[F], query_rand: list[F], joint_rand: list[F], num_shares: int) -> list[F]` is a verifier's query algorithm. This is used to "query" the measurement and proof. The result of the query (i.e., the output of this function) is called the "verifier message". In addition to the measurement and proof, this algorithm takes as input the query randomness `query_rand` and the joint randomness `joint_rand`. The former is used only by the verifier. Here, num_shares specifies how many shares were generated.
+* `flp.truncate(meas: list[F]) -> list[F]` maps an encoded measurement (e.g., the bit-encoding of the measurement) to an aggregatable output (e.g., the singleton vector containing the measurement). The length of the input MUST be `MEAS_LEN` and the length of the output MUST be `OUTPUT_LEN`.
```

Comment:
I don't think it's necessary to list functions for APIs we use. For example, it's sufficient to say something like "Mastic is defined in terms of the `Flp` interface specified in {{Section XXX of !VDAF}}. Likewise for `Field` here.

Reply:
We are listing all the functions we use so the description is self contained (most are also in the !VDAF).

Reply:
I don't think the document needs to be completely self-contained. For example, we don't describe how `Flp` is implemented, i.e., how the ZKPs are constructed.

Moreover I think it's counter-productive if we try to make things self-contained. {{!VDAF}} is not yet complete and is subject to change, so the less we try to copy and paste from there, the less work we have to do to keep them in-sync.

Reply:
Well, we are addressing a TODO that says _"List all functions we use in alphabetical order."_ We just need to define what is "all functions". I think every function that we explicitly use in `draft-mouris-cfrg-mastic.md` satisfies this definition (unless a function is defined elsewhere in the text, like `gen` for example). Otherwise, where do you draw the line of what is copied here vs. referenced in !VDAF. I am open to alternative views of what "all functions" should be here.

Reply:
I was thinking a subset of the functions in [this list](https://cfrg.github.io/draft-irtf-cfrg-vdaf/draft-irtf-cfrg-vdaf.html#section-2-11). Note that the TODO says "{{Section 2 of !VDAF}}", i.e., refers to the conventions in {{!VDAF}}.

Reply:
I guess it is a bit unclear what are the criteria for inclusion in [this list](https://cfrg.github.io/draft-irtf-cfrg-vdaf/draft-irtf-cfrg-vdaf.html#section-2-11). For example, it can be any function that needs to be imported from `common.py`. Or is it more subjective (i.e., _here are some functions that we often use_)?

In our case, I think it is reasonable to list the imported functions that are referenced in our text (along with basic functionalities). As a reader, I find it useful to have these definitions in one place (but personal preferences can differ on this). If it makes more sense, we can split the list to basic functions and APIs.

Reply:
> I guess it is a bit unclear what are the criteria for inclusion in [this list](https://cfrg.github.io/draft-irtf-cfrg-vdaf/draft-irtf-cfrg-vdaf.html#section-2-11). For example, it can be any function that needs to be imported from `common.py`. Or is it more subjective (i.e., _here are some functions that we often use_)?

This is the complete list of basic functions that the specifications of Prio3 and Poplar1 use, but we don't explicitly define in full detail. What they do _should_ be clear from the description; if anything is unclear, let me know :)

> In our case, I think it is reasonable to list the imported functions that are referenced in our text (along with basic functionalities). As a reader, I find it useful to have these definitions in one place (but personal preferences can differ on this). If it makes more sense, we can split the list to basic functions and APIs.

At a minimum, we should split the basic functions and APIs into separate lists. If we're going to re-define the APIs (by this I mean the `Field`, `Xof`, and `Flp` classes defined in {{!VDAF}}), then they also need to be more complete then they are currently. Each API is consists not just of instance methods, e.g., `flp.encode()`, but potentially class methods or constructors, e.g., `XofTurboShake128()`. They also define constants, e.g., `flp.MEAS_LEN` and have generic parameters, e.g., `F`.

But again, I don't think we should be re-defining APIs. Not only does this make our jobs harder in the short term, I don't think it helps much with an RFC. These documents are not meant to be self-contained as long as dependencies have clear refernces.